Now better:

>AST fuzzer (debug, actions) — Task failed: $?=210

Hm, some issue in `getaddrinfo`

<details>

```
2021.12.16 08:45:41.274358 [ 230 ] {} <Trace> BaseDaemon: Received signal 5
2021.12.16 08:45:41.274778 [ 234 ] {} <Fatal> BaseDaemon: ########################################
2021.12.16 08:45:41.274872 [ 234 ] {} <Fatal> BaseDaemon: (version 21.13.1.538, build id: 2D910F2D2502C654) (from thread 99) (no query) Received signal Trace/breakpoint trap (5)
2021.12.16 08:45:41.274933 [ 234 ] {} <Fatal> BaseDaemon: 
2021.12.16 08:45:41.275024 [ 234 ] {} <Fatal> BaseDaemon: Stack trace: 0x7ff1a47f108f 0x7ff1a46f58b8 0x7ff1a47f05fa 0x7ff1a46f4861 0x7ff1a46f58b8 0x7ff1a46f5983 0x7ff1a46f4995 0x7ff1a46d862c 0x7ff1a46d8ed9 0x7ff1a46994f5 0x7ff1a469b0d9 0x27d23d34 0x27d24f16 0x27cf0203 0x27cf0158 0x157e1982 0x27d79645 0x157d8956 0x27d9a2d3 0x157d67e7 0x156f4e65 0x7ff1a45b90b3 0x156f4a6e
2021.12.16 08:45:41.275106 [ 234 ] {} <Fatal> BaseDaemon: 4. ? @ 0x7ff1a47f108f in ?
2021.12.16 08:45:41.275153 [ 234 ] {} <Fatal> BaseDaemon: 5. _dl_catch_exception @ 0x7ff1a46f58b8 in ?
2021.12.16 08:45:41.275189 [ 234 ] {} <Fatal> BaseDaemon: 6. ? @ 0x7ff1a47f05fa in ?
2021.12.16 08:45:41.275226 [ 234 ] {} <Fatal> BaseDaemon: 7. ? @ 0x7ff1a46f4861 in ?
2021.12.16 08:45:41.275264 [ 234 ] {} <Fatal> BaseDaemon: 8. _dl_catch_exception @ 0x7ff1a46f58b8 in ?
2021.12.16 08:45:41.275302 [ 234 ] {} <Fatal> BaseDaemon: 9. _dl_catch_error @ 0x7ff1a46f5983 in ?
2021.12.16 08:45:41.275340 [ 234 ] {} <Fatal> BaseDaemon: 10. __libc_dlopen_mode @ 0x7ff1a46f4995 in ?
2021.12.16 08:45:41.275395 [ 234 ] {} <Fatal> BaseDaemon: 11. ? @ 0x7ff1a46d862c in ?
2021.12.16 08:45:41.275433 [ 234 ] {} <Fatal> BaseDaemon: 12. __nss_lookup_function @ 0x7ff1a46d8ed9 in ?
2021.12.16 08:45:41.275470 [ 234 ] {} <Fatal> BaseDaemon: 13. ? @ 0x7ff1a46994f5 in ?
2021.12.16 08:45:41.275507 [ 234 ] {} <Fatal> BaseDaemon: 14. getaddrinfo @ 0x7ff1a469b0d9 in ?
2021.12.16 08:45:41.362055 [ 234 ] {} <Fatal> BaseDaemon: 15. ./obj-x86_64-linux-gnu/../contrib/poco/Net/src/DNS.cpp:79: Poco::Net::DNS::hostByName(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, unsigned int) @ 0x27d23d34 in /workspace/clickhouse
2021.12.16 08:45:41.445293 [ 234 ] {} <Fatal> BaseDaemon: 16. ./obj-x86_64-linux-gnu/../contrib/poco/Net/src/DNS.cpp:194: Poco::Net::DNS::thisHost() @ 0x27d24f16 in /workspace/clickhouse
2021.12.16 08:45:41.517369 [ 234 ] {} <Fatal> BaseDaemon: 17. ./obj-x86_64-linux-gnu/../base/base/getFQDNOrHostName.cpp:11: (anonymous namespace)::getFQDNOrHostNameImpl() @ 0x27cf0203 in /workspace/clickhouse
2021.12.16 08:45:41.578008 [ 234 ] {} <Fatal> BaseDaemon: 18. ./obj-x86_64-linux-gnu/../base/base/getFQDNOrHostName.cpp:23: getFQDNOrHostName() @ 0x27cf0158 in /workspace/clickhouse
2021.12.16 08:45:41.670984 [ 234 ] {} <Fatal> BaseDaemon: 19. ./obj-x86_64-linux-gnu/../programs/server/Server.cpp:769: DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) @ 0x157e1982 in /workspace/clickhouse
2021.12.16 08:45:41.763793 [ 234 ] {} <Fatal> BaseDaemon: 20. ./obj-x86_64-linux-gnu/../contrib/poco/Util/src/Application.cpp:334: Poco::Util::Application::run() @ 0x27d79645 in /workspace/clickhouse
2021.12.16 08:45:41.900908 [ 234 ] {} <Fatal> BaseDaemon: 21. ./obj-x86_64-linux-gnu/../programs/server/Server.cpp:401: DB::Server::run() @ 0x157d8956 in /workspace/clickhouse
2021.12.16 08:45:41.978903 [ 234 ] {} <Fatal> BaseDaemon: 22. ./obj-x86_64-linux-gnu/../contrib/poco/Util/src/ServerApplication.cpp:611: Poco::Util::ServerApplication::run(int, char**) @ 0x27d9a2d3 in /workspace/clickhouse
2021.12.16 08:45:42.115231 [ 234 ] {} <Fatal> BaseDaemon: 23. ./obj-x86_64-linux-gnu/../programs/server/Server.cpp:179: mainEntryClickHouseServer(int, char**) @ 0x157d67e7 in /workspace/clickhouse
2021.12.16 08:45:42.132032 [ 234 ] {} <Fatal> BaseDaemon: 24. ./obj-x86_64-linux-gnu/../programs/main.cpp:378: main @ 0x156f4e65 in /workspace/clickhouse
2021.12.16 08:45:42.132085 [ 234 ] {} <Fatal> BaseDaemon: 25. __libc_start_main @ 0x7ff1a45b90b3 in ?
2021.12.16 08:45:46.044067 [ 234 ] {} <Fatal> BaseDaemon: 26. _start @ 0x156f4a6e in /workspace/clickhouse
2021.12.16 08:45:47.156760 [ 234 ] {} <Fatal> BaseDaemon: Calculated checksum of the binary: 27B40AB4F18BD9CC69E2165982CA0BF4. There is no information about the reference checksum.
2021.12.16 08:45:47.156830 [ 234 ] {} <Information> SentryWriter: Not sending crash report
```

</details>
>Integration tests (asan, actions) [2/3] — fail: 1, passed: 546, flaky: 0

- test_storage_kerberized_kafka/test.py::test_kafka_json_as_string_no_kdc - #kafak

>Integration tests (asan, actions) [3/3] — fail: 1, passed: 626, flaky: 1

- test_s3_zero_copy_replication/test.py::test_s3_zero_copy_replication[s3]

>Integration tests (release, actions) [2/2] — fail: 1, passed: 929, flaky: 1

- test_merge_tree_blob_storage/test.py::test_freeze_unfreeze

```
E           helpers.client.QueryRuntimeException: Client failed! Return code: 233, stderr: Received exception from server (version 21.13.1):
E           Code: 1001. DB::Exception: Received from 172.16.12.3:9000. DB::Exception: Azure::Storage::StorageException: The specified blob does not exist.
```

--- a/src/Common/ThreadStatus.cpp
+++ b/src/Common/ThreadStatus.cpp
@@ -72,6 +72,24 @@ static thread_local bool has_alt_stack = false;
 #endif
 
 
+std::vector<ThreadGroupStatus::ProfileEventsCountersAndMemory> ThreadGroupStatus::getProfileEventsCountersAndMemoryForThreads()
+{
+    std::lock_guard guard(mutex);
+
+    /// It is OK to move it, since it is enough to report statistics for the thread at least once.
+    auto stats = std::move(finished_threads_counters_memory);
+    for (auto * thread : threads)
+    {
+        stats.emplace_back(ProfileEventsCountersAndMemory{
+            thread->performance_counters.getPartiallyAtomicSnapshot(),
+            thread->memory_tracker.get(),
+            thread->thread_id,
+        });
+    }
+
+    return stats;
+}
+
 ThreadStatus::ThreadStatus()
     : thread_id{getThreadId()}
 {
@@ -145,6 +163,11 @@ ThreadStatus::~ThreadStatus()
     if (thread_group)
     {
         std::lock_guard guard(thread_group->mutex);
+        thread_group->finished_threads_counters_memory.emplace_back(ThreadGroupStatus::ProfileEventsCountersAndMemory{
+            performance_counters.getPartiallyAtomicSnapshot(),
+            memory_tracker.get(),
+            thread_id,
+        });
         thread_group->threads.erase(this);
     }
 

--- a/src/Common/ThreadStatus.cpp
+++ b/src/Common/ThreadStatus.cpp
@@ -139,6 +139,7 @@ ThreadStatus::~ThreadStatus()
     {
         /// It's a minor tracked memory leak here (not the memory itself but it's counter).
         /// We've already allocated a little bit more than the limit and cannot track it in the thread memory tracker or its parent.
+        tryLogCurrentException(log);
     }
 
     if (thread_group)

--- a/src/Common/ThreadStatus.h
+++ b/src/Common/ThreadStatus.h
@@ -61,6 +61,13 @@ using ThreadStatusPtr = ThreadStatus *;
 class ThreadGroupStatus
 {
 public:
+    struct ProfileEventsCountersAndMemory
+    {
+        ProfileEvents::Counters::Snapshot counters;
+        Int64 memory_usage;
+        UInt64 thread_id;
+    };
+
     mutable std::mutex mutex;
 
     ProfileEvents::Counters performance_counters{VariableContext::Process};
@@ -83,6 +90,10 @@ public:
 
     String query;
     UInt64 normalized_query_hash = 0;
+
+    std::vector<ProfileEventsCountersAndMemory> finished_threads_counters_memory;
+
+    std::vector<ProfileEventsCountersAndMemory> getProfileEventsCountersAndMemoryForThreads();
 };
 
 using ThreadGroupStatusPtr = std::shared_ptr<ThreadGroupStatus>;

--- a/src/Server/TCPHandler.cpp
+++ b/src/Server/TCPHandler.cpp
@@ -947,28 +947,27 @@ void TCPHandler::sendProfileEvents()
     ThreadIdToCountersSnapshot new_snapshots;
     ProfileEventsSnapshot group_snapshot;
     {
-        std::lock_guard guard(thread_group->mutex);
-        snapshots.reserve(thread_group->threads.size());
-        for (auto * thread : thread_group->threads)
+        auto stats = thread_group->getProfileEventsCountersAndMemoryForThreads();
+        snapshots.reserve(stats.size());
+
+        for (auto & stat : stats)
         {
-            auto const thread_id = thread->thread_id;
+            auto const thread_id = stat.thread_id;
             if (thread_id == current_thread_id)
                 continue;
             auto current_time = time(nullptr);
-            auto counters = thread->performance_counters.getPartiallyAtomicSnapshot();
-            auto memory_usage = thread->memory_tracker.get();
             auto previous_snapshot = last_sent_snapshots.find(thread_id);
             auto increment =
                 previous_snapshot != last_sent_snapshots.end()
-                ? CountersIncrement(counters, previous_snapshot->second)
-                : CountersIncrement(counters);
+                ? CountersIncrement(stat.counters, previous_snapshot->second)
+                : CountersIncrement(stat.counters);
             snapshots.push_back(ProfileEventsSnapshot{
                 thread_id,
                 std::move(increment),
-                memory_usage,
+                stat.memory_usage,
                 current_time
             });
-            new_snapshots[thread_id] = std::move(counters);
+            new_snapshots[thread_id] = std::move(stat.counters);
         }
 
         group_snapshot.thread_id    = 0;

--- a/tests/queries/0_stateless/02050_client_profile_events.reference
+++ b/tests/queries/0_stateless/02050_client_profile_events.reference
@@ -1,5 +1,5 @@
 0
 100000
-SelectedRows: 131010 (increment)
+[ 0 ] SelectedRows: 131010 (increment)
 OK
 OK

--- a/tests/queries/0_stateless/02050_client_profile_events.sh
+++ b/tests/queries/0_stateless/02050_client_profile_events.sh
@@ -7,7 +7,7 @@ CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
 # do not print any ProfileEvents packets
 $CLICKHOUSE_CLIENT -q 'select * from numbers(1e5) format Null' |& grep -c 'SelectedRows'
 # print only last (and also number of rows to provide more info in case of failures)
-$CLICKHOUSE_CLIENT --print-profile-events --profile-events-delay-ms=-1 -q 'select * from numbers(1e5)' 2> >(grep -o -e 'SelectedRows: .*$' -e Exception) 1> >(wc -l)
+$CLICKHOUSE_CLIENT --print-profile-events --profile-events-delay-ms=-1 -q 'select * from numbers(1e5)' 2> >(grep -o -e '\[ 0 \] SelectedRows: .*$' -e Exception) 1> >(wc -l)
 # print everything
 profile_events="$($CLICKHOUSE_CLIENT --max_block_size 1 --print-profile-events -q 'select sleep(1) from numbers(2) format Null' |& grep -c 'SelectedRows')"
 test "$profile_events" -gt 1 && echo OK || echo "FAIL ($profile_events)"

